lightning/src/offers/signer.rs

```diff
@@ -0,0 +1,172 @@
+	Derived(MetadataMaterial),
+
+	/// Metadata and signing pubkey to be derived from message contents and given material.
+	DerivedSigningPubkey(MetadataMaterial),
+}
+
+impl Metadata {
+	pub fn as_bytes(&self) -> Option<&Vec<u8>> {
+		match self {
+			Metadata::Bytes(bytes) => Some(bytes),
+			Metadata::Derived(material) => None,
+			Metadata::DerivedSigningPubkey(material) => None,
+		}
+	}
+
+	pub fn material(&self) -> Option<&MetadataMaterial> {
+		match self {
+			Metadata::Bytes(_) => None,
+			Metadata::Derived(material) => Some(material),
+			Metadata::DerivedSigningPubkey(material) => Some(material),
+		}
+	}
+
+	pub fn material_mut(&mut self) -> Option<&mut MetadataMaterial> {
+		match self {
+			Metadata::Bytes(_) => None,
+			Metadata::Derived(material) => Some(material),
+			Metadata::DerivedSigningPubkey(material) => Some(material),
+		}
+	}
+
+	pub fn into_parts(self) -> (Vec<u8>, Option<KeyPair>) {
+		match self {
+			Metadata::Bytes(bytes) => (bytes, None),
+			Metadata::Derived(metadata_material) => (metadata_material.into_metadata(), None),
+			Metadata::DerivedSigningPubkey(metadata_material) => {
+				let (metadata, keys) = metadata_material.into_parts();
+				(metadata, Some(keys))
+			},
+		}
+	}
+}
+
+impl fmt::Debug for Metadata {
+	fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+		match self {
+			Metadata::Bytes(bytes) => bytes.fmt(f),
+			Metadata::Derived(material) => f.write_str("Derived"),
+			Metadata::DerivedSigningPubkey(material) => f.write_str("DerivedSigningPubkey"),
+		}
+	}
+}
+
+impl PartialEq for Metadata {
+	fn eq(&self, other: &Self) -> bool {
+		match self {
+			Metadata::Bytes(bytes) => if let Metadata::Bytes(other_bytes) = other {
+				bytes == other_bytes
+			} else {
+				false
+			},
+			Metadata::Derived(material) => false,
+			Metadata::DerivedSigningPubkey(material) => false,
+		}
+	}
+}
+
+/// Material used to create metadata for a message. Once initialized, write the applicable data from
+/// the message into it and call [`Metadata::into_parts`] to retrieve the metadata and derived
+/// signing pubkey, if any.
+#[derive(Clone)]
+pub(super) struct MetadataMaterial {
+	nonce: EncryptedNonce,
+	hmac: HmacEngine<Sha256>,
+}
+
+impl MetadataMaterial {
+	pub fn new(nonce: Nonce, expanded_key: &ExpandedKey) -> Self {
+		Self {
+			nonce: expanded_key.encrypt_nonce_for_offer(nonce),
+			hmac: expanded_key.hmac_for_offer(nonce),
+		}
+	}
+
+	fn into_metadata(self) -> Vec<u8> {
+		let mut bytes = self.nonce.as_slice().to_vec();
+		bytes.extend_from_slice(&Hmac::from_engine(self.hmac).into_inner());
+		bytes
+	}
+
+	fn into_parts(self) -> (Vec<u8>, KeyPair) {
+		let secp_ctx = Secp256k1::new();
+		let hmac = Hmac::from_engine(self.hmac);
+		let privkey = SecretKey::from_slice(hmac.as_inner()).unwrap();
+		let keys = KeyPair::from_secret_key(&secp_ctx, &privkey);
+		(self.nonce.as_slice().to_vec(), keys)
+	}
+}
+
+impl io::Write for MetadataMaterial {
```

Comment:
Gonna noodle on this a bit. Currently, `into_parts` returns `(Vec<u8>, Option<KeyPair>)`, but then we immediately go and make a `Metadata::Bytes` with the `Vec<u8>`. We probably should just convert it internally to a `Metadata::Bytes` instead of consuming now that `Metadata` is used throughout. Then we'd need to return the `KeyPair` if applicable.

Reply:
Cleaned this up quite a bit based on feedback. Now there is only a `derive_from` method taking a `Writeable` TLV stream.